### PR TITLE
executor: show operators' concurrency information in results of `EXPLAIN`

### PR DESCRIPTION
**Pull request by Reminiscent**
### What problem does this PR solve? 
Add the operator's concurrency information in `execution info` when we use the `EXPLAIN` statement.

After this PR, you can see:
```
mysql> explain select avg(t1.a) from t1, t2 where t1.b = t2.b group by t1.a;
+------------------------------+-------+------+------------------------------------------------------------------------------------------------+
| id                           | count | task | operator info                                                                                  |
+------------------------------+-------+------+------------------------------------------------------------------------------------------------+
| HashAgg_11                   | 3.75  | root | group by:Column#9, funcs:avg(Column#0), HashAggPartialConcurrency:4, HashAggFinalConcurrency:4 |
| └─Projection_20              | 3.75  | root | cast(Column#1), Column#1, ProjectionConcurrency:0                                              |
|   └─HashRightJoin_13         | 3.75  | root | inner join, inner:TableReader_16, equal:[eq(Column#5, Column#2)], HashJoinConcurrency:5        |
|     ├─TableReader_16         | 3.00  | root | data:Selection_15                                                                              |
|     │ └─Selection_15         | 3.00  | cop  | not(isnull(Column#5))                                                                          |
|     │   └─TableScan_14       | 3.00  | cop  | table:t2, range:[-inf,+inf], keep order:false, stats:pseudo                                    |
|     └─TableReader_19         | 5.99  | root | data:Selection_18                                                                              |
|       └─Selection_18         | 5.99  | cop  | not(isnull(Column#2))                                                                          |
|         └─TableScan_17       | 6.00  | cop  | table:t1, range:[-inf,+inf], keep order:false, stats:pseudo                                    |
+------------------------------+-------+------+------------------------------------------------------------------------------------------------+



```

### What is changed and how it works?
Some operators including projection/ hashAgg/ hashJoin/ IndexNestedLoopHashJoin , can be executed concurrently.  We add their concurrency information to the result when we execute the "explain analyze" statement. Besides, if the root task can't execute concurrently, the concurrency number will be zero.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: util/execdetails/execdetails.go [hunk 1]
```diff
@@ -381,13 +383,19 @@ func (e *RuntimeStats) Record(d time.Duration, rowNum int) {
...
 }
 
+// SetConcurrencyNum sets the concurrency num.
+func (e *RuntimeStats) SetConcurrencyNum(concurrencyNum int64) {
+	atomic.StoreInt64(&e.concurrency, concurrencyNum)
+}
+
 func (e *RuntimeStats) String() string {
-	return fmt.Sprintf("time:%v, loops:%d, rows:%d", time.Duration(e.consume), e.loop, e.rows)
+	return fmt.Sprintf("time:%v, loops:%d, rows:%d, concurrency:%d", time.Duration(e.consume), e.loop, e.rows, e.concurrency)
```
Comment (zz-jason) [inline util/execdetails/execdetails.go]: It's better display concurrency details for each operator. For example, in hash aggregate operator, we can show how many partial and final workers respectively.

File: util/execdetails/execdetails.go [hunk 2]
```diff
@@ -381,13 +383,19 @@ func (e *RuntimeStats) Record(d time.Duration, rowNum int) {
 	atomic.AddInt32(&e.loop, 1)
 	atomic.AddInt64(&e.consume, int64(d))
 	atomic.AddInt64(&e.rows, int64(rowNum))
+	atomic.AddInt64(&e.concurrency, 0)
```
Comment (qw4990) [inline util/execdetails/execdetails.go]: Is `1` better than `0` in this case?

Comment (Reminiscent) [reply util/execdetails/execdetails.go]: In the original version, 0 means non-parallel execution, such as https://github.com/pingcap/tidb/blob/master/executor/projection.go#L88 . I'm not sure which is better.

File: util/execdetails/execdetails.go (hunk 1 above)
Comment (Reminiscent) [reply util/execdetails/execdetails.go]: Update. PTAL

**Comment by Reminiscent** [comment]
> Will we support show the time consumption of each worker in explain analyze?

I think we can implement this feature in the next PR. What's your opinion?

**Comment by Reminiscent** [comment]
> How about add another field to display the concurrency information? For example, `concurrency`.

I think it's better to put the concurrency information in the `operator info` field. Because only few operators need to show the concurrency information. It can save some storage. What's your opinion? @zz-jason

File: planner/core/explain.go [hunk 3]
```diff
@@ -161,7 +161,8 @@ func (p *PhysicalSelection) ExplainInfo() string {
 
 // ExplainInfo implements PhysicalPlan interface.
 func (p *PhysicalProjection) ExplainInfo() string {
-	return string(expression.ExplainExpressionList(p.Exprs))
+	return string(expression.ExplainExpressionList(p.Exprs)) +
+		fmt.Sprintf(", ProjectionConcurrency:%d", p.ProjectionConcurrency)
```
Comment (zz-jason) [inline planner/core/explain.go]: How about this, it's shorter:
```suggestion
		fmt.Sprintf(", concurrency:%d", p.ProjectionConcurrency)
```

BTW, if concurrency is `0`, it's no need to print it out.

File: planner/core/explain.go [hunk 4]
```diff
@@ -209,6 +210,13 @@ func (p *basePhysicalAgg) ExplainInfo() string {
 	return buffer.String()
 }
 
+// ExplainInfo implements PhysicalPlan interface.
+func (p *PhysicalHashAgg) ExplainInfo() string {
+	return p.basePhysicalAgg.ExplainInfo() + fmt.Sprintf(
+		", HashAggPartialConcurrency:%d, HashAggFinalConcurrency:%d",
```
Comment (zz-jason) [inline planner/core/explain.go]:
```suggestion
		", PartialConcurrency:%d, FinalConcurrency:%d",
```

File: planner/core/explain.go [hunk 5]
```diff
@@ -233,6 +241,9 @@ func (p *PhysicalIndexJoin) ExplainInfo() string {
 		fmt.Fprintf(buffer, ", other cond:%s",
 			expression.SortedExplainExpressionList(p.OtherConditions))
 	}
+	if p.IndexLookupJoinConcurrency > 0 {
+		fmt.Fprintf(buffer, ", IndexLookupJoinConcurrency:%d", p.IndexLookupJoinConcurrency)
```
Comment (zz-jason) [inline planner/core/explain.go]:
```suggestion
		fmt.Fprintf(buffer, ", Concurrency:%d", p.IndexLookupJoinConcurrency)
```

File: planner/core/explain.go [hunk 6]
```diff
@@ -260,6 +271,9 @@ func (p *PhysicalHashJoin) ExplainInfo() string {
 		fmt.Fprintf(buffer, ", other cond:%s",
 			expression.SortedExplainExpressionList(p.OtherConditions))
 	}
+	if p.Concurrency > 0 {
+		fmt.Fprintf(buffer, ", HashJoinConcurrency:%d", p.Concurrency)
```
Comment (zz-jason) [inline planner/core/explain.go]:
```suggestion
		fmt.Fprintf(buffer, ", Concurrency:%d", p.Concurrency)
```

**Pull request closed without merging**